compiler/qsc_circuit/src/builder.rs

```diff
@@ -370,6 +373,92 @@ impl Builder {
     }
 }
 
+/// Provides support for qubit id allocation, measurement and
+/// reset operations for Base Profile targets.
+///
+/// Since qubit reuse is disallowed, a mapping is maintained
+/// from allocated qubit ids to hardware qubit ids. Each time
+/// a qubit is reset, it is remapped to a fresh hardware qubit.
+///
+/// Note that even though qubit reset & reuse is disallowed,
+/// qubit ids are still reused for new allocations.
+/// Measurements are tracked and deferred.
+#[derive(Default)]
+struct Remapper {
+    next_meas_id: usize,
+    next_qubit_id: usize,
+    next_qubit_hardware_id: HardwareId,
+    qubit_map: IndexMap<usize, HardwareId>,
+    measurements: Vec<(HardwareId, usize)>,
```

Comment:
This isn't a public API, can we create a type for this to improve clarity/intent?

Reply:
changing to an explicit type made me realize the list of tuples was only ever being used as a count of measurements per qubit (used to be more when this struct was part of QIR codegen, but not it's just for circuits). Simplified to an `IndexMap` of hardware id to count of measurements.